akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1104,3 +1105,55 @@ private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraph
   }
   override def toString = "Reduce"
 }
+
+/**
+ * INTERNAL API
+ */
+final class RecoverWith[T, M](pf: PartialFunction[Throwable, Graph[SourceShape[T], M]]) extends SimpleLinearGraphStage[T] {
+  override def initialAttributes = DefaultAttributes.recoverWith
+
+  override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+    var available: Option[SubSinkInlet[T]] = None
+    var finished = false
+
+    def pushOut(): Unit = {
+      val src = available.get
+      push(out, src.grab())
+      if (!src.isClosed) src.pull()
+      else completeStage()
+    }
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = push(out, grab(in))
+      override def onUpstreamFailure(ex: Throwable) = if (pf.isDefinedAt(ex)) switchTo(pf(ex)) else failStage(ex)
+    })
+
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = pull(in)
+    })
+
+    val outHandler = new OutHandler {
+      override def onPull(): Unit = if (available.nonEmpty && isAvailable(out)) pushOut()
+    }
+
+    def switchTo(source: Graph[SourceShape[T], M]): Unit = {
+      val sinkIn = new SubSinkInlet[T]("RecoverWithSink")
```

Comment:
Wouldn't it be better perhaps to not have the `available` option, but put everything regarding the switched mode in this scope, where it can be directly referenced, as it is accessed for every element passing through?